### PR TITLE
Homework13

### PR DESCRIPTION
**Pull request by gl-alexander**
Adds backwards compatibility between scene types by storing constant albedo, as well as a texture name in CRTMaterial (the overhead is small enough to be negligable, as std::string only takes up 16 - 32 bytes of storage if uninitialized and materials are generally fewer)

Fixes refraction sphere bug, removes the black dot in the middle (issue was unnormalized vertex normal)

Adds an additional image for the sphere scene with Schlick's approximation in the Fresnel formula, gives a nice reflective effect

Adds optimizations to the renderer. Several test functions included, like renderByRegions, renderByBuckets and the most optimized renderWithAABB. The renderWithAABB function uses the Bucket method and the AABB check when raytracing

The renderByBuckets makes use of the CRTBucketQueue class, which is thread safe with a mutex for accessing. boolean return value tells us if we've emptied the queue, so the threads in CRTRenderer can stop

The test values for the function don't seem quite right, will be checking them before uploading the final results, otherwise the duration for rendering is in the Source_HW13.cpp file under each test function

### CODE REVIEW

**Comment by gl-alexander** [comment]
AABB creation had a small mistake, the results look more believable now, with the benchmarks added to the Images/Homework13/results.png image

**Comment by gl-alexander** [comment]
Refactors the textures as discussed, 
Replaces the dynamic polymorphism with std::variant, which was the fastest approach out of the ones I tried. With this small changes to the way textures are parsed and stored were added
Another optimization is storing a pointer to the texture in the actual CRTMaterial, as opposed to the texture name. This improves performance for more complex scenes

The benchmarks with the Dynamic Polymorphism approach (in Release build with a single thread) on Homework12/scene4.crtscene
avg.: 0.7028 s
After introducing std::variant and storing a pointer to the texture
avg.: 0.6531 s 

The improvement is small, but keeping in mind the simplicity of the scene, the ~8% faster time is worth it

File: SourceCode/CRTRenderer.cpp [hunk 1]
```diff
@@ -152,10 +176,17 @@ CRTVector CRTRenderer::shadeRefractive(const CRTRay& ray, const Intersection& da
     CRTVector reflectedDirection = reflect(ray.direction, normalVector);
     CRTRay reflectedRay{ data.hitPoint + normalVector * REFLECTION_BIAS, reflectedDirection, RayType::REFLECTIVE, ray.depth + 1 };
 
-    float fresnel = 0.5f * pow(1.0f + dotPr, 5);
+    // Improved Fresnel calculation using Schlick's approximation
```
Comment (VladislavVulchevChaos) [inline SourceCode/CRTRenderer.cpp:179]: Nice :)

File: Images/Homework13/results_hw13.png
Comment (VladislavVulchevChaos) [inline Images/Homework13/results_hw13.png:1]: Very very good results

File: SourceCode/Scene/CRTBucketQueue.cpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+{
+	unsigned width = scene->getSettings().imageSettings.width;
+	unsigned height = scene->getSettings().imageSettings.height;
+	unsigned bucketSize = scene->getSettings().imageSettings.bucketSize;
+
+	unsigned bucketsPerWidth = width / bucketSize;
+	unsigned bucketsPerHeight = height / bucketSize;
+
+	unsigned lastBucketWidth = bucketSize + (width % bucketSize);
+	unsigned lastBucketHeight = bucketSize + (height % bucketSize); // to avoid leaving out pixels at the edges
```
Comment (VladislavVulchevChaos) [inline SourceCode/Scene/CRTBucketQueue.cpp:13]: Good addition

**Pull request merged**